src/Elm/Declare.elm

```diff
@@ -64,159 +65,284 @@ In that case you can do something like this using `callFrom`:
 
 -}
 
-import Elm exposing (Declaration, Expression)
+import Elm exposing (Expression)
 import Elm.Annotation
+import Elm.Arg
 import Elm.Syntax.Expression as Exp
 import Elm.Syntax.ModuleName as ModuleName
 import Internal.Compiler as Compiler
 import Internal.Format as Format
 
 
+type alias Module val =
+    { name : List String
+    , declarations : List Elm.Declaration
+    , call : val
+    }
+
+
+{-| -}
+elmModule : List String -> val -> Module val
+elmModule name call =
+    { name = name
+    , call = call
+    , declarations = []
+    }
+
+
+{-| -}
+with : Decl required -> Module (required -> val) -> Module val
+with decl mod =
```

Comment:
We also need `withUnexposed : Decl optional -> Module val -> Module val` I think.

Plus either have `with` implicitly expose or add an `expose/exposeWith : Decl -> Decl`